modules/new_serverpod_auth/serverpod_auth_jwt/serverpod_auth_jwt_server/lib/src/business/authentication_tokens.dart

```diff
@@ -0,0 +1,263 @@
+        level: LogLevel.debug,
+        exception: e,
+        stackTrace: stackTrace,
+      );
+
+      return null;
+    } catch (e) {
+      return null;
+    }
+  }
+
+  /// Creates a new token pair for a given auth user.
+  ///
+  /// This is akin to creating a new session, and should be used after a successful login or registration.
+  static Future<TokenPair> createTokens(
+    final Session session, {
+    required final UuidValue authUserId,
+    required final Set<Scope> scopes,
+
+    /// Extra claims to be added to the JWT.
+    ///
+    /// These are added on the top level of the paylaod, so be sure not to conflict with the [registered claims](https://datatracker.ietf.org/doc/html/rfc7519#section-4.1),
+    /// as those will just overwrite any custom claims given here.
+    ///
+    /// Since these claims will be embedded in every access token (also across rotations), one should be mindful about there length and total size.
+    final Map<String, dynamic>? extraClaims,
+    final Transaction? transaction,
+  }) async {
+    final secret = _generateRefreshTokenSecret();
+    final newHash = RefreshTokenSecretHash.createHash(secret: secret);
+
+    final refreshToken = await RefreshToken.db.insertRow(
+      session,
+      RefreshToken(
+        authUserId: authUserId,
+        fixedSecret: ByteData.sublistView(_generateRefreshTokenFixedSecret()),
+        rotatingSecretHash: ByteData.sublistView(newHash.hash),
+        rotatingSecretSalt: ByteData.sublistView(newHash.salt),
+        scopeNames: scopes.names,
+        extraClaims: extraClaims != null ? jsonEncode(extraClaims) : null,
+      ),
+      transaction: transaction,
+    );
+
+    return TokenPair(
+      refreshToken: RefreshTokenString.buildRefreshTokenString(
+        refreshToken: refreshToken,
+        rotatingSecret: secret,
+      ),
+      accessToken: JwtUtil.createJwt(refreshToken),
+    );
+  }
+
+  /// Returns a new refresh & access token pair.
+  ///
+  /// This invalidates the previous refresh token.
+  static Future<TokenPair> rotateRefreshToken(
+    final Session session, {
+    required final String refreshToken,
+    final Transaction? transaction,
+  }) async {
+    final RefreshTokenStringData refreshTokenData;
+
+    try {
+      refreshTokenData = RefreshTokenString.parseRefreshTokenString(
+        refreshToken,
+      );
+    } catch (e, stackTrace) {
+      session.log(
+        'Received malformed refresh token',
+        exception: e,
+        stackTrace: stackTrace,
+        level: LogLevel.debug,
+      );
+
+      throw RefreshTokenMalformedException();
+    }
+
+    var refreshTokenRow = await RefreshToken.db.findById(
+      session,
+      refreshTokenData.id,
+      transaction: transaction,
+    );
+
+    if (refreshTokenRow == null ||
+        !uint8ListAreEqual(
+          Uint8List.sublistView(refreshTokenRow.fixedSecret),
+          refreshTokenData.fixedSecret,
+        )) {
+      throw RefreshTokenNotFoundException();
```

Comment:
```suggestion
      // If the `fixedSecret` does not match, we do not delete the refresh token.
      // (Since the `id` is encoded on every access token and thus might be known to 3rd parties or leak, even after the access token itself has expired.)
      throw RefreshTokenNotFoundException();
```